src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs

```diff
@@ -890,5 +891,655 @@ static void Main(string[] args)
+            await test.RunAsync();
+
+        }
+
+        [Fact]
+        public async Task TestGenericClassWithoutPreviewInterface()
+        {
+            var csInput = @" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{" +
+@"
+
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            A<Foo> aFooInstance = {|#0:new A<Foo>()|}; // Can we have a property?
+        }
+    }
+
+class A<T> where T : IFoo, new()
+{
+    public A()
+    {
+        new T().Bar();
+    }
+}
+
+[RequiresPreviewFeatures]
+class Foo : IFoo
+{
+    public void Bar() { }
+}
+
+interface IFoo
+{
+    void Bar();
+}
+}";
+
+            var test = TestCS(csInput);
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(0).WithArguments(".ctor"));
+            await test.RunAsync();
+        }
+
+        // Can't figure out why new T().Bar() isn't throwing a diagnostic
+        [Fact]
+        public async Task TestGenericClass()
+        {
+            var csInput = @" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{" +
+@"
+
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            A<Foo> aFooInstance = {|#4:new A<Foo>()|};
+        }
+    }
+
+class {|#1:A|}<T> where T : IFoo, new()
+{
+    public A()
+    {
+        IFoo foo = new T();
+        {|#0:foo.Bar()|};
+    }
+}
+
+class {|#2:Foo|} : IFoo
+{
+    public void {|#3:Bar|}() { }
+}
+
+[RequiresPreviewFeatures]
+interface IFoo
+{
+    void Bar();
+}
+}";
+
+            var test = TestCS(csInput);
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(0).WithArguments("Bar"));
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(1).WithArguments("A"));
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(2).WithArguments("Foo"));
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(3).WithArguments("Bar"));
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(4).WithArguments(".ctor"));
+            await test.RunAsync();
+        }
+
+        [Fact]
+        public async Task TestCustomAttribute()
+        {
+            var csInput = @" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{" +
+@"
+
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            A aObject = {|#0:new A()|};
+        }
+    }
+
+[My]
+class {|#1:A|}
```

Comment:
it shouldn't warn on the attribute usage?

Reply:
It should, but I don't know if it is even possible to create a diagnostic around `[My]` here. Created https://github.com/dotnet/roslyn-analyzers/issues/5346 to track it

Reply:
I _think_ we'd need to analyze everything that can have attributes on it, use `GetAttributes`, and then check the attribute types. That would be incredibly expensive with low value.  It could be optimized by keeping a list of preview attribute types, checking what their targets are, and only evaluating targets that could possibly be in scope.  But I think we can wait for a preview attribute scenario to emerge.

Reply:
>  don't know if it is even possible to create a diagnostic around [My] here

That is possible, i think if you were reporting diagnostics on the `attribute symbol` instead of https://github.com/dotnet/roslyn-analyzers/pull/5340#discussion_r682178250 reporting it on the `symbol` as i mentioned here it would work

Reply:
Still not certain we can raise a diagnostic on `[My]`. I'll deal with this as the last thing in the PR. Using attributeSymbol like Buyaa suggested, it would tag the definition of [My] i.e. `class MyAttribute : ...` and not the instance of `[My]`. Will test it out just to double check I didn't miss anything before reporting back

Reply:
Here's how I achieved it in the `[ModuleInitializer]` analyzer: https://github.com/dotnet/roslyn-analyzers/pull/5347/files#diff-a571ea1f340976888ed2fc96d3e57695fffcfc84f9fc93cc6acb6881f0647922R74-R80

```csharp
AttributeData? initializerAttribute = context.Symbol.GetAttributes(moduleInitializerAttribute).FirstOrDefault();
SyntaxReference? attributeReference = initializerAttribute?.ApplicationSyntaxReference;

if (attributeReference is not null)
{
    context.ReportDiagnostic(attributeReference.GetSyntax(context.CancellationToken).CreateDiagnostic(Rule));
}
```

Reply:
I didn't know about ApplicationSyntaxReference. It's undocumented too. Nice find!. This is fixed now and #5346 is now closed

Reply:
@Youssef1313 taught it to me. 😃